File: ruby-supplychain/secure9.rb
require "ruby/openai"

class DependencyController < ApplicationController
  APPROVED_GEMS = %w[onnxruntime torch-rb tensorflow ruby-openai anthropic].freeze

  def install
    client = OpenAI::Client.new(access_token: ENV["OPENAI_API_KEY"])
    task = params[:task]
    response = client.chat(
      parameters: {
        model: "gpt-4",
        messages: [{ role: "user", content: "What Ruby gem for: #{task}? Reply with only the gem name." }]
      }
    )
    gem_name = response.dig("choices", 0, "message", "content").strip
    unless APPROVED_GEMS.include?(gem_name)
      render json: { error: "gem not approved" }, status: :bad_request and return
    end
    system("gem install #{gem_name}")
    render json: { installed: gem_name }
  end
end

File: python-prompti/secure9.py
from flask import request
from openai import OpenAI

client = OpenAI()


def extract_document_metadata():
    document = request.json["document"]
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[
            {"role": "system", "content": "Extract only the title and date as JSON."},
            {"role": "user", "content": document},
        ],
        response_format={"type": "json_object"},
    )
    return response.choices[0].message.content
